### PR TITLE
[AutoTVM] Use PopenPool in XGBoostCostModel

### PR DESCRIPTION
**Pull request by shingjan**
This PR intends to replace the use of `multiprocessing.pool` with `PopenPoolExecutor` in `AutoTVM.XgBoostCostModel`

@vinx13 @junrushao1994 @tqchen

### CODE REVIEW

File: python/tvm/autotvm/tuner/xgboost_cost_model.py [hunk 1]
```diff
@@ -327,14 +329,9 @@ def _get_feature(self, indexes):
...
-            # If we are forking, we can pass arguments in globals for better performance
-            if multiprocessing.get_start_method(False) == "fork":
-                feas = pool.map(self.feature_extract_func, need_extract)
-            else:
-                args = [(self.space.get(x), self.target, self.task) for x in need_extract]
-                feas = pool.map(self.feature_extract_func, args)
+            feas = pool.map_with_error_catching(self.feature_extract_func, need_extract)
             for i, fea in zip(need_extract, feas):
-                fea_cache[i] = fea
+                fea_cache[i] = fea.value
```
Comment (vinx13) [inline python/tvm/autotvm/tuner/xgboost_cost_model.py:334]:
```suggestion
                fea_cache[i] = fea.value if fea.status == StatusKind.COMPLETE else None
```
let's be cautious with potential failure in rare cases

**Review by vinx13 — COMMENTED** [review]
could you share some performance numbers?

File: python/tvm/autotvm/tuner/xgboost_cost_model.py [hunk 2]
```diff
@@ -144,6 +146,12 @@ def __init__(
         self._sample_size = 0
         self._reset_pool(self.space, self.target, self.task)
 
+    def _initializer(space, target, task):
```
Comment (vinx13) [inline python/tvm/autotvm/tuner/xgboost_cost_model.py:149]: decorate it with `@staticmethod`

Comment (shingjan) [reply python/tvm/autotvm/tuner/xgboost_cost_model.py:149]: I am moving this out of `XGBoostCostModel` class for a linting issue. Right now this function would be defined at line 351.

File: python/tvm/autotvm/tuner/xgboost_cost_model.py (hunk 1 above)
Comment (shingjan) [reply python/tvm/autotvm/tuner/xgboost_cost_model.py:334]: done

**Comment by shingjan** [comment]
Performance benchmarks on Kraken:

Running `tune_conv2d_cude.py` in `tutorials` with `n_trial` set to `100`

| SA elapsed time | Fork      | PopenPool |
| --- | --- | ---|
| 1st run| 69.85 | 67.68 |
| 2nd run| 66.34     | 68.75       |
| 3rd run| 70.10   | 67.58        |

File: python/tvm/autotvm/tuner/xgboost_cost_model.py [hunk 3]
```diff
@@ -247,13 +242,14 @@ def fit_log(self, records, plan_size, min_seed_records=500):
...
             raise RuntimeError("Invalid feature type: " + self.fea_type)
-        res = pool.map(feature_extract_func, data)
+        result = pool.map_with_error_catching(feature_extract_func, data)
 
         # filter out feature with different shapes
         fea_len = len(self._get_feature([0])[0])
 
         xs, ys = [], []
-        for x, y in res:
+        for res in result:
```
Comment (vinx13) [inline python/tvm/autotvm/tuner/xgboost_cost_model.py:251]:
```suggestion
        for res in result:
            if res.status != StatusKind.COMPLETE:
                continue
```

Comment (shingjan) [reply python/tvm/autotvm/tuner/xgboost_cost_model.py:251]: done

**Review by vinx13 — APPROVED** [review]

**Comment by junrushao1994** [comment]
Thanks a lot @shingjan @vinx13! It is merged!

**Pull request merged**